File: agent/linux/pkg/reputation/ReputationLookup.go
package reputation

import (
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"io"
	"log"
	"net/http"
	"os"
	"sync"
	"time"
)

type Verdict string

const (
	VerdictClean     Verdict = "Clean"
	VerdictMalicious Verdict = "Malicious"
	VerdictUnknown   Verdict = "Unknown"
)

type ReputationResult struct {
	SHA256     string    `json:"sha256"`
	Verdict    Verdict   `json:"verdict"`
	ThreatName string    `json:"threatName,omitempty"`
	Confidence int       `json:"confidence"` // 0-100
	CachedAt   time.Time `json:"cachedAt"`
}

type ReputationLookup struct {
	backendURL string
	cache      map[string]ReputationResult
	mu         sync.RWMutex
	httpClient *http.Client
}

func NewReputationLookup(backendURL string) *ReputationLookup {
	return &ReputationLookup{
		backendURL: backendURL,
		cache:      make(map[string]ReputationResult),
		httpClient: &http.Client{Timeout: 8 * time.Second},
	}
}

// LookupFile hashes a file and checks its reputation.
func (r *ReputationLookup) LookupFile(filePath string) (*ReputationResult, error) {
	hash, err := sha256File(filePath)
	if err != nil {
		return nil, fmt.Errorf("hash file: %w", err)
	}
	return r.LookupHash(hash)
}

// LookupHash checks a SHA-256 hash against the backend reputation service.
func (r *ReputationLookup) LookupHash(hash string) (*ReputationResult, error) {
	// Check local cache first
	r.mu.RLock()
	if cached, ok := r.cache[hash]; ok && time.Since(cached.CachedAt) < 10*time.Minute {
		r.mu.RUnlock()
		return &cached, nil
	}
	r.mu.RUnlock()

	// Query backend
	url := fmt.Sprintf("%s/api/v1/reputation/%s", r.backendURL, hash)
	resp, err := r.httpClient.Get(url)
	if err != nil {
		log.Printf("[ReputationLookup] Backend unreachable, returning Unknown for %s: %v", hash[:16], err)
		result := ReputationResult{SHA256: hash, Verdict: VerdictUnknown, CachedAt: time.Now()}
		return &result, nil
	}
	defer resp.Body.Close()

	var result ReputationResult
	if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
		return nil, fmt.Errorf("decode response: %w", err)
	}
	result.CachedAt = time.Now()

	r.mu.Lock()
	r.cache[hash] = result
	r.mu.Unlock()

	return &result, nil
}

func sha256File(path string) (string, error) {
	f, err := os.Open(path)
	if err != nil {
		return "", err
	}
	defer f.Close()
	h := sha256.New()
	if _, err := io.Copy(h, f); err != nil {
		return "", err
	}
	return hex.EncodeToString(h.Sum(nil)), nil
}

File: agent/macos/Sources/FenrirAgent/ReputationLookup.swift
import Foundation
import CryptoKit

enum ReputationVerdict: String, Codable {
    case clean = "Clean"
    case malicious = "Malicious"
    case unknown = "Unknown"
}

struct ReputationResult: Codable {
    let sha256: String
    let verdict: ReputationVerdict
    let threatName: String?
    let confidence: Int
}

class ReputationLookup {
    private let config: AgentConfig
    private var cache: [String: (result: ReputationResult, cachedAt: Date)] = [:]
    private let lock = NSLock()
    private let session: URLSession
    private let cacheTTL: TimeInterval = 600 // 10 minutes

    init(config: AgentConfig) {
        self.config = config
        let cfg = URLSessionConfiguration.default
        cfg.timeoutIntervalForRequest = 8
        self.session = URLSession(configuration: cfg)
    }

    func lookupFile(at path: String, completion: @escaping (ReputationResult) -> Void) {
        DispatchQueue.global(qos: .userInitiated).async { [weak self] in
            guard let self = self else { return }
            guard let hash = self.sha256(filePath: path) else {
                completion(ReputationResult(sha256: "unknown", verdict: .unknown, threatName: nil, confidence: 0))
                return
            }
            self.lookupHash(hash, completion: completion)
        }
    }

    func lookupHash(_ hash: String, completion: @escaping (ReputationResult) -> Void) {
        // Check cache
        lock.lock()
        if let cached = cache[hash], Date().timeIntervalSince(cached.cachedAt) < cacheTTL {
            let result = cached.result
            lock.unlock()
            completion(result)
            return
        }
        lock.unlock()

        // Query backend reputation endpoint
        let reputationBase = config.backendURL.deletingLastPathComponent()
        guard let url = URL(string: "\(reputationBase)/api/v1/reputation/\(hash)") else {
            completion(ReputationResult(sha256: hash, verdict: .unknown, threatName: nil, confidence: 0))
            return
        }

        let task = session.dataTask(with: url) { [weak self] data, _, error in
            guard let self = self, let data = data, error == nil,
                  let result = try? JSONDecoder().decode(ReputationResult.self, from: data) else {
                completion(ReputationResult(sha256: hash, verdict: .unknown, threatName: nil, confidence: 0))
                return
            }
            self.lock.lock()
            self.cache[hash] = (result: result, cachedAt: Date())
            self.lock.unlock()
            completion(result)
        }
        task.resume()
    }

    private func sha256(filePath: String) -> String? {
        guard let data = try? Data(contentsOf: URL(fileURLWithPath: filePath)) else { return nil }
        let digest = SHA256.hash(data: data)
        return digest.map { String(format: "%02x", $0) }.joined()
    }
}
